### PR TITLE
[Happy] Enable happy test in pre-submit checks

### LINKED ISSUE #2591: [HAPPY] Add parallel test engine for Happy
Currently happy tests are running sequentially, in automake system, it has parallel test engine to trigger multiple parallel tests, when switching to gn build, we also need to run happy tests in parallel.

### PR DESCRIPTION
**Pull request by erjiaqing**
#### Problem
* Currently, Happy test framework was added to CHIP, but no tests are enabled.

#### Summary of Changes
* Follow-ups after #2319 is merged
* Wrapper for running happy tests in gn/ninja
* `happy_test` template
* Enables them in pre-submit checks
* ~~Run all unit tests in gn_test~~
* Happy tests can be invoked by 
  ```
  scripts/build/gn_gen.sh
  RUN_HAPPY_TESTS=1 scripts/tests/gn_tests.sh
  ```
* Move all unit tests to seperate github actions (`unit_test`).
* Logs are uploads to artifacts in github actions

fixes #2591

### CODE REVIEW

File: gn/chip/chip_test_standalone.gni [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import("//build_overrides/chip.gni")
+
+import("${chip_root}/gn/chip/tests.gni")
+
+assert(chip_build_tests)
+
+template("chip_test_standalone") {
```
Comment (mspang) [inline gn/chip/chip_test_standalone.gni:21]: I'm not spotting what the difference is between chip_test_standalone and chip_test_suite. Why did we need to duplicate it?

File: src/test_driver/happy/tests/standalone/inet/BUILD.gn [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import("//build_overrides/chip.gni")
+
+import("${chip_root}/gn/chip/happy_test.gni")
+
+happy_test("inet_tests") {
+    deps = [
```
Comment (mspang) [inline src/test_driver/happy/tests/standalone/inet/BUILD.gn:20]: data_deps?

**Review by mspang — CHANGES_REQUESTED** [review]

File: src/test_driver/happy/tests/standalone/inet/BUILD.gn (hunk 2 above)
Comment (mspang) [reply src/test_driver/happy/tests/standalone/inet/BUILD.gn:20]: Also prefer to depend on only the targets actually used, not the whole group

data_deps = [
  "${chip_root}/src/inet/tests:TestInetLayerMulticast"
]

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -37,14 +37,26 @@ jobs:
             image: connectedhomeip/chip-build:0.4.7
             volumes:
                 - "/tmp/log_output:/tmp/test_logs"
+                - "/tmp/happy_test_logs:/tmp/happy_test_logs"
+            options: "--privileged"
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply .github/workflows/build.yaml:41]: ip net namespace interaction need priviledge capability for docker, happy is lightweight ip namespace orchestration platform.

File: scripts/tests/happy_tests.sh [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    echo "Bootstrapping Happy Test"
+    set -e
+    # Bootstrap Happy
+    cd "$REPO_DIR/third_party/happy/repo"
+
+    # Override happy log dir config.
+    python3 -c 'import json, os; file = open("happy/conf/main_config.json"); data = json.load(file); data["log_directory"] = os.environ["HAPPY_LOG_DIR"]; out = open("happy/conf/main_config.json", "w"); json.dump(data, out);'
+
+    apt-get update
+    apt-get install -y bridge-utils \
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply scripts/tests/happy_tests.sh:41]: agree, this need to put in dockerfile, push it to the docerhub

File: scripts/tests/happy_test_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+parser.add_argument('--test-bin-dir', dest='bin_dir', type=str, nargs='?', default='.',
+                    help='The path of test binaries')
+parser.add_argument('--ci', dest='ci', type=bool, nargs='?', default=False,
+                    help='Set this if running script under venv but happy is installed globally')
+
+if __name__ == '__main__':
+    args = parser.parse_args()
+    print(args, file=sys.stderr)
+    print(test_environ, file=sys.stderr)
+    # GN will run Python in venv, which will break happy test
```
[the remark this answers is not in the document]
Comment (erjiaqing) [reply scripts/tests/happy_test_wrapper.py:41]: happy is not installed in venv

**Review by woody-apple — APPROVED** [review]

File: scripts/tests/happy_test_wrapper.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mspang) [reply scripts/tests/happy_test_wrapper.py:41]: we can install it there?

**Review by rwalker-apple — CHANGES_REQUESTED** [review]
Please use the PR template.  Please describe the problem this solves, i.e. the motivation for making this change.

File: scripts/tests/happy_tests.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (andy31415) [reply scripts/tests/happy_tests.sh]: Could you create a PR to update docker (version and packages)?
I can handle the push once a PR is submitted (and even a temporary build to test before submitting, since version will be bumped)

[the remark this answers is not in the document]
Comment (erjiaqing) [reply scripts/tests/happy_tests.sh]: For now, we decided to use separate job for tests, since we don't want to introduce sudos and privileged containers in main build.

File: src/inet/tests/BUILD.gn [hunk 6]
```diff
@@ -37,8 +37,6 @@ chip_test_suite("tests") {
     "TestInetLayer.cpp",
     "TestInetLayer.h",
     "TestInetLayerCommon.cpp",
-    "TestInetLayerDNS.cpp",
```
Comment (rwalker-apple) [inline src/inet/tests/BUILD.gn]: why are these removed?

Comment (erjiaqing) [reply src/inet/tests/BUILD.gn]: This is not removed, but was moved to line 67. Also this is a merge commit, the original change was made on master in other prs.

**Comment by erjiaqing** [comment]
@yunhanw-google PTAL

File: src/inet/tests/BUILD.gn [hunk 7]
```diff
@@ -55,3 +56,28 @@ chip_test_suite("tests") {
     "TestInetErrorStr",
   ]
 }
+
+
+chip_test_standalone("standalone_tests") {
+  public_configs = [ ":tests_config" ]
+
+  sources = [
+    "TapAddrAutoconf.cpp",
```
[the remark this answers is not in the document]
Comment (erjiaqing) [reply src/inet/tests/BUILD.gn]: I tried to do so, but seems it will break the build. Added as TODO

File: gn/chip/chip_test_standalone.gni (hunk 1 above)
Comment (erjiaqing) [reply gn/chip/chip_test_standalone.gni]: Done, chip_test_suite requires a Driver.cpp, but TestInetLayerMulticast does not need a Driver.cpp, modified the test to use chip_test_suite.

File: scripts/tests/happy_tests.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply scripts/tests/happy_tests.sh:44]: the host machine in github action don't have ipv6 support. we need to run it inside docker with ipv6 configuration. happy orchestrate lightweight ip netns, which need privildege support in docker. 

Agree. please consider to add these apt-get packages in docker as @andreilitvin suggesed.

File: scripts/tests/happy_test_wrapper.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply scripts/tests/happy_test_wrapper.py:41]: happy can be run in virtualenv, there may be one bug in recent happy change.

**Review by yunhanw-google — APPROVED** [review]

**Review by rwalker-apple — APPROVED** [review]

File: scripts/tests/happy_tests.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (mspang) [reply scripts/tests/happy_tests.sh:44]: Something like this: https://github.com/project-chip/connectedhomeip/pull/2854

File: src/inet/tests/BUILD.gn [hunk 8]
```diff
@@ -63,9 +65,39 @@ chip_test_suite("tests") {
...
       "TestInetLayerDNS.cpp",
-      "TestInetLayerMulticast.cpp",
       "TestLwIPDNS.cpp",
     ]
     tests += [ "TestInetLayerDNS" ]
   }
 }
+
+# The following binaries should be executed by happy.
+chip_test_suite("happy_tests") {
```
[the remark this answers is not in the document]
Comment (erjiaqing) [reply src/inet/tests/BUILD.gn:87]: Fixed

File: scripts/tests/happy_test_wrapper.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (erjiaqing) [reply scripts/tests/happy_test_wrapper.py:41]: Did some attempts to get happy available in venv, but failed, one of the reason is failed to install psutil in venv, see https://github.com/erjiaqing/connectedhomeip/runs/1174544304 for details. I think this can be a follow up.

**Review by BroderickCarlin — APPROVED** [review]

**Comment by mspang** [comment]
I've run this successfully now and understand why we have the 2 python environments (it's a bug).

My biggest gripe with it is that this runs "sudo" inside ninja when running the tests locally. This is not only risky, but it doesn't work unless sudo was run recently on the same tty, or is configured for passwordless invocation. When we go to run the tests, it's likely to block waiting for password entry.

I think it is really problematic to require root access to run tests locally. Could we solve this by running all the happy & cirque tests in KVM? That would solve the other issues we're having with various environments (incl. docker) where we need to add special kernel modules. It's also a much safer and a well supported path (see Android emulator).

If we check these in before addressing this, I think they the happy test groups should be totally removed by default and there should be a big warning about the risks & caveats before turning them on and trying to run them locally.

**Comment by yunhanw-google** [comment]
It seems currently github action don't support nested virtualization.
https://github.com/actions/virtual-environments/issues/183

Also in Weave, Happy has run for pretty long time for presubmit, the risky is relatively low. What is the detailed risk for sudo ninja?

> I've run this successfully now and understand why we have the 2 python environments (it's a bug).
> 
> My biggest gripe with it is that this runs "sudo" inside ninja when running the tests locally. This is not only risky, but it doesn't work unless sudo was run recently on the same tty, or is configured for passwordless invocation. When we go to run the tests, it's likely to block waiting for password entry.
> 
> I think it is really problematic to require root access to run tests locally. Could we solve this by running all the happy & cirque tests in KVM? That would solve the other issues we're having with various environments (incl. docker) where we need to add special kernel modules. It's also a much safer and well supported path (see Android emulator).
> 
> If we check these in before addressing this, I think the happy test groups should be totally removed by default and there should be a big warning about the risks & caveats before developers try to run them locally.

**Comment by mspang** [comment]
> It seems currently github action don't support nested virtualization.
> [actions/virtual-environments#183](https://github.com/actions/virtual-environments/issues/183)
> 
> Also in Weave, Happy has run for pretty long time for presubmit, the risky is relatively low. What is the detailed risk for sudo ninja?

The risks is that the host is damaged by the tests, since they run with dangerous privileges.

`sudo ninja` will result in pretty immediate problems, e.g. root-owned build outputs which need sudo to delete. Building as root is pretty universally discouraged in industry (see e.g. fakeroot).

Other solutions to problems like this:
- user namespaces
- setuid sandbox helpers. I believe this is largely obsoleted in favor of user namespaces though.

Has happy investigated any of these options?

Example:
```
  # as regular user with user namespaces support
  unshare --map-root-user -n -m
  mount --make-private /
  mount -t tmpfs tmpfs /run
  ip netns add test
  ip netns exec test ip link
  ip tuntap add mode tap tap0
  ip netns exec test ip tuntap add mode tap tap1
  ip link add veth0 type veth peer name veth1
  ip link set veth1 netns test
  ip link
  ip netns exec test ip link
```

Output:
```
1: lo: <LOOPBACK> mtu 65536 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: tap0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether 22:3a:1b:39:3d:9a brd ff:ff:ff:ff:ff:ff
4: veth0@if3: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether 0a:8e:5b:9f:96:25 brd ff:ff:ff:ff:ff:ff link-netns test
```

Output [test]:

```
1: lo: <LOOPBACK> mtu 65536 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: tap0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether 22:3a:1b:39:3d:9a brd ff:ff:ff:ff:ff:ff
4: veth0@if3: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether 0a:8e:5b:9f:96:25 brd ff:ff:ff:ff:ff:ff link-netns test
```

None of that requires root outside the user namespace. What happy operations require true root?

> 
> > I've run this successfully now and understand why we have the 2 python environments (it's a bug).
> > My biggest gripe with it is that this runs "sudo" inside ninja when running the tests locally. This is not only risky, but it doesn't work unless sudo was run recently on the same tty, or is configured for passwordless invocation. When we go to run the tests, it's likely to block waiting for password entry.
> > I think it is really problematic to require root access to run tests locally. Could we solve this by running all the happy & cirque tests in KVM? That would solve the other issues we're having with various environments (incl. docker) where we need to add special kernel modules. It's also a much safer and well supported path (see Android emulator).
> > If we check these in before addressing this, I think the happy test groups should be totally removed by default and there should be a big warning about the risks & caveats before developers try to run them locally.

**Comment by yunhanw-google** [comment]
Hi, Michael
Thanks for the comments. I think we can run happy under user namespaces in docker as below using unshare --map-root-user , however, the priviledge is still needed since ip netns need priviledge in docker. I think unshare would be the way moving forward on this PR.

newuser@47b3efdae915:/happy/topologies$ unshare --map-root-user -n -m
# mount --make-private /
  mount -t tmpfs tmpfs /run# 
# whoami
root
# happy-network-add HomeThread thread
# happy-network-address HomeThread 2001:db8:111:1::
# happy-network-add HomeWiFi wifi
# happy-network-address HomeWiFi 2001:db8:222:2::
# happy-network-address HomeWiFi 10.0.1.0
# happy-node-add ThreadNode
# happy-node-join ThreadNode HomeThread
happy-node-add BorderRouter
# # happy-node-join BorderRouter HomeThread
# happy-node-join BorderRouter HomeWiFi

# happy-state

State Name:  happy

NETWORKS   Name         Type   State                                     Prefixes
     HomeThread       thread      UP                       2001:0db8:0111:0001/64 

       HomeWiFi         wifi      UP                                    10.0.1/24 
                                                           2001:0db8:0222:0002/64 


NODES      Name    Interface    Type                                          IPs
   BorderRouter        wlan0    wifi                                  10.0.1.2/24 
                                       2001:0db8:0222:0002:88ab:feff:fe5f:cff9/64 

                       wpan0  thread   2001:0db8:0111:0001:98c1:ebff:fe9a:de10/64 

     ThreadNode        wpan0  thread   2001:0db8:0111:0001:84db:54ff:feb8:6532/64 

          onhub

**Comment by mspang** [comment]
> Hi, Michael
> Thanks for the comments. I think we can run happy under user namespaces in docker as below using unshare --map-root-user , however, the priviledge is still needed since ip netns need priviledge in docker. I think unshare would be the way moving forward on this PR.

Hm, looks like the default docker seccomp BPF policy blocks unshare. We could make our own policy but for now using --privileged seems OK to achieve that. GitHub Actions runs on Hyper-V and I'm not worried about that as much.

For builds outside docker, +1 for using user namespaces. We can document that this needs to be turned on for the happy tests. It's much safer than running as root.

> 
> newuser@47b3efdae915:/happy/topologies$ unshare --map-root-user -n -m
> 
> > mount --make-private /
> > mount -t tmpfs tmpfs /run#
> > whoami
> > root
> > happy-network-add HomeThread thread
> > happy-network-address HomeThread 2001:db8:111:1::
> > happy-network-add HomeWiFi wifi
> > happy-network-address HomeWiFi 2001:db8:222:2::
> > happy-network-address HomeWiFi 10.0.1.0
> > happy-node-add ThreadNode
> > happy-node-join ThreadNode HomeThread
> > happy-node-add BorderRouter
> > happy-node-join BorderRouter HomeThread
> > happy-node-join BorderRouter HomeWiFi
> > happy-state
> > State Name:  happy
> > NETWORKS   Name         Type   State                                     Prefixes
> > HomeThread       thread      UP                       2001:0db8:0111:0001/64
> > ```
> >    HomeWiFi         wifi      UP                                    10.0.1/24 
> >                                                        2001:0db8:0222:0002/64 
> > ```
> > 
> > 
> > NODES      Name    Interface    Type                                          IPs
> > BorderRouter        wlan0    wifi                                  10.0.1.2/24
> > 2001:0db8:0222:0002:88ab:feff:fe5f:cff9/64
> > ```
> >                    wpan0  thread   2001:0db8:0111:0001:98c1:ebff:fe9a:de10/64 
> > 
> >  ThreadNode        wpan0  thread   2001:0db8:0111:0001:84db:54ff:feb8:6532/64 
> > 
> >       onhub
> > ```



> Hi, Michael
> Thanks for the comments. I think we can run happy under user namespaces in docker as below using unshare --map-root-user , however, the priviledge is still needed since ip netns need priviledge in docker. I think unshare would be the way moving forward on this PR.
> 
> newuser@47b3efdae915:/happy/topologies$ unshare --map-root-user -n -m
> 
> > mount --make-private /
> > mount -t tmpfs tmpfs /run#
> > whoami
> > root
> > happy-network-add HomeThread thread
> > happy-network-address HomeThread 2001:db8:111:1::
> > happy-network-add HomeWiFi wifi
> > happy-network-address HomeWiFi 2001:db8:222:2::
> > happy-network-address HomeWiFi 10.0.1.0
> > happy-node-add ThreadNode
> > happy-node-join ThreadNode HomeThread
> > happy-node-add BorderRouter
> > happy-node-join BorderRouter HomeThread
> > happy-node-join BorderRouter HomeWiFi
> > happy-state
> > State Name:  happy
> > NETWORKS   Name         Type   State                                     Prefixes
> > HomeThread       thread      UP                       2001:0db8:0111:0001/64
> > ```
> >    HomeWiFi         wifi      UP                                    10.0.1/24 
> >                                                        2001:0db8:0222:0002/64 
> > ```
> > 
> > 
> > NODES      Name    Interface    Type                                          IPs
> > BorderRouter        wlan0    wifi                                  10.0.1.2/24
> > 2001:0db8:0222:0002:88ab:feff:fe5f:cff9/64
> > ```
> >                    wpan0  thread   2001:0db8:0111:0001:98c1:ebff:fe9a:de10/64 
> > 
> >  ThreadNode        wpan0  thread   2001:0db8:0111:0001:84db:54ff:feb8:6532/64 
> > 
> >       onhub
> > ```

**Comment by mspang** [comment]
> > Hi, Michael
> > Thanks for the comments. I think we can run happy under user namespaces in docker as below using unshare --map-root-user , however, the priviledge is still needed since ip netns need priviledge in docker. I think unshare would be the way moving forward on this PR.
> 
> Hm, looks like the default docker seccomp BPF policy blocks unshare. We could make our own policy but for now using --privileged seems OK to achieve that. GitHub Actions runs on Hyper-V and I'm not worried about that as much.
> 
> For builds outside docker, +1 for using user namespaces. We can document that this needs to be turned on for the happy tests. It's much safer than running as root.

I spent a while digging into why we can't unshare in docker. Turns out BPF isn't the only (or hardest) issue; procfs & sysfs masking also interferes with remounting sysfs, which ip-netns needs to do to enter a new namespace. Probably will be eventually fixed in order to enable nested rootless containers. Until then using a privileged container is the only way to get nesting working, unfortunately.

> 
> > newuser@47b3efdae915:/happy/topologies$ unshare --map-root-user -n -m
> > > mount --make-private /
> > > mount -t tmpfs tmpfs /run#
> > > whoami
> > > root
> > > happy-network-add HomeThread thread
> > > happy-network-address HomeThread 2001:db8:111:1::
> > > happy-network-add HomeWiFi wifi
> > > happy-network-address HomeWiFi 2001:db8:222:2::
> > > happy-network-address HomeWiFi 10.0.1.0
> > > happy-node-add ThreadNode
> > > happy-node-join ThreadNode HomeThread
> > > happy-node-add BorderRouter
> > > happy-node-join BorderRouter HomeThread
> > > happy-node-join BorderRouter HomeWiFi
> > > happy-state
> > > State Name:  happy
> > > NETWORKS   Name         Type   State                                     Prefixes
> > > HomeThread       thread      UP                       2001:0db8:0111:0001/64
> > > ```
> > >    HomeWiFi         wifi      UP                                    10.0.1/24 
> > >                                                        2001:0db8:0222:0002/64 
> > > ```
> > > 
> > > 
> > > NODES      Name    Interface    Type                                          IPs
> > > BorderRouter        wlan0    wifi                                  10.0.1.2/24
> > > 2001:0db8:0222:0002:88ab:feff:fe5f:cff9/64
> > > ```
> > >                    wpan0  thread   2001:0db8:0111:0001:98c1:ebff:fe9a:de10/64 
> > > 
> > >  ThreadNode        wpan0  thread   2001:0db8:0111:0001:84db:54ff:feb8:6532/64 
> > > 
> > >       onhub
> > > ```
> 
> > Hi, Michael
> > Thanks for the comments. I think we can run happy under user namespaces in docker as below using unshare --map-root-user , however, the priviledge is still needed since ip netns need priviledge in docker. I think unshare would be the way moving forward on this PR.
> > newuser@47b3efdae915:/happy/topologies$ unshare --map-root-user -n -m
> > > mount --make-private /
> > > mount -t tmpfs tmpfs /run#
> > > whoami
> > > root
> > > happy-network-add HomeThread thread
> > > happy-network-address HomeThread 2001:db8:111:1::
> > > happy-network-add HomeWiFi wifi
> > > happy-network-address HomeWiFi 2001:db8:222:2::
> > > happy-network-address HomeWiFi 10.0.1.0
> > > happy-node-add ThreadNode
> > > happy-node-join ThreadNode HomeThread
> > > happy-node-add BorderRouter
> > > happy-node-join BorderRouter HomeThread
> > > happy-node-join BorderRouter HomeWiFi
> > > happy-state
> > > State Name:  happy
> > > NETWORKS   Name         Type   State                                     Prefixes
> > > HomeThread       thread      UP                       2001:0db8:0111:0001/64
> > > ```
> > >    HomeWiFi         wifi      UP                                    10.0.1/24 
> > >                                                        2001:0db8:0222:0002/64 
> > > ```
> > > 
> > > 
> > > NODES      Name    Interface    Type                                          IPs
> > > BorderRouter        wlan0    wifi                                  10.0.1.2/24
> > > 2001:0db8:0222:0002:88ab:feff:fe5f:cff9/64
> > > ```
> > >                    wpan0  thread   2001:0db8:0111:0001:98c1:ebff:fe9a:de10/64 
> > > 
> > >  ThreadNode        wpan0  thread   2001:0db8:0111:0001:84db:54ff:feb8:6532/64 
> > > 
> > >       onhub
> > > ```

**Review by mspang — CHANGES_REQUESTED** [review]
Marking request changes for removal of root access requirement and use of `sudo` when running the tests via ninja. Please ping me if next steps are not clear.

**Comment by erjiaqing** [comment]
> Marking request changes for removal of root access requirement and use of `sudo` when running the tests via ninja. Please ping me if next steps are not clear.

I checked that happy tests are running directly in VM and ninja is not executed under sudo?

**Comment by erjiaqing** [comment]
> Marking request changes for removal of root access requirement and use of `sudo` when running the tests via ninja. Please ping me if next steps are not clear.

PR Updated, found sudo is not available when using user namespace, added a face_sudo.sh as a workaround.

**Comment by mspang** [comment]
> > Marking request changes for removal of root access requirement and use of `sudo` when running the tests via ninja. Please ping me if next steps are not clear.
> 
> PR Updated, found sudo is not available when using user namespace, added a face_sudo.sh as a workaround.

Why is this required?

From happy source:
```
    def getRunAsRootPrefixList(self):
        if "SUDO" in list(os.environ.keys()):
            return [os.environ["SUDO"]]
        elif os.getuid() == 0:
            # We are already root
            return []
        else:
            return ["sudo"]
```

The `getuid()` check should work out, right?

File: src/inet/tests/BUILD.gn [hunk 9]
```diff
@@ -23,19 +23,12 @@ config("tests_config") {
   include_dirs = [ "." ]
 }
 
-chip_test_suite("tests") {
-  output_name = "libInetLayerTests"
-
+static_library("chip_inet_test_common_source") {
```
Comment (mspang) [inline src/inet/tests/BUILD.gn:26]: I think this will break nRF connect & ESP32 tests.

Shouldn't need to do change this, the tests are already a static library, see below comment.

File: src/inet/tests/BUILD.gn [hunk 10]
```diff
@@ -63,9 +77,26 @@ chip_test_suite("tests") {
...
+# The following binaries should be executed by happy.
+chip_test_suite("happy_tests") {
+  output_name = "libHappyTestInetCommon"
+
+  sources = [ "TestInetLayerMulticast.cpp" ]
+
+  public_configs = [ ":tests_config" ]
+
+  public_deps = [
+    ":chip_inet_test_common_source",
```
Comment (mspang) [inline src/inet/tests/BUILD.gn:95]: Change this to

`  ":tests_common"`

and merge `chip_inet_test_common_source` back into `tests`.

I think this should resolve breaking issue with ESP32 tests.

File: src/inet/tests/Makefile.am [hunk 11]
```diff
@@ -0,0 +1,252 @@
+#
```
Comment (mspang) [inline src/inet/tests/Makefile.am:1]: Delete this file.

**Review by mspang — COMMENTED** [review]
Thanks, just a few more fixes..

**Comment by erjiaqing** [comment]
> > > Marking request changes for removal of root access requirement and use of `sudo` when running the tests via ninja. Please ping me if next steps are not clear.
> > 
> > 
> > PR Updated, found sudo is not available when using user namespace, added a face_sudo.sh as a workaround.
> 
> Why is this required?
> 
> From happy source:
> 
> ```
>     def getRunAsRootPrefixList(self):
>         if "SUDO" in list(os.environ.keys()):
>             return [os.environ["SUDO"]]
>         elif os.getuid() == 0:
>             # We are already root
>             return []
>         else:
>             return ["sudo"]
> ```
> 
> The `getuid()` check should work out, right?

In happy:
```python
    def getRunAsUserPrefixList(self, username=None):
        if username is None:
            username = getpass.getuser()

        if "SUDO" in list(os.environ.keys()):
            return [os.environ["SUDO"], "-u", username]
        else:
            return ["sudo", "-u", username]
```

sudo is required here.

File: src/BUILD.gn [hunk 12]
```diff
@@ -63,4 +63,12 @@ if (chip_build_tests) {
       deps += [ "${chip_root}/src/lib/shell/tests" ]
     }
   }
+
+  if (current_os != "zephyr") {
```
Comment (mspang) [inline src/BUILD.gn:67]: `current_os == "linux"` right?

Comment (erjiaqing) [reply src/BUILD.gn]: You are right, happy only supports linux.

**Comment by mspang** [comment]
> > > > Marking request changes for removal of root access requirement and use of `sudo` when running the tests via ninja. Please ping me if next steps are not clear.
> > > 
> > > 
> > > PR Updated, found sudo is not available when using user namespace, added a face_sudo.sh as a workaround.
> > 
> > 
> > Why is this required?
> > From happy source:
> > ```
> >     def getRunAsRootPrefixList(self):
> >         if "SUDO" in list(os.environ.keys()):
> >             return [os.environ["SUDO"]]
> >         elif os.getuid() == 0:
> >             # We are already root
> >             return []
> >         else:
> >             return ["sudo"]
> > ```
> > 
> > 
> > The `getuid()` check should work out, right?
> 
> In happy:
> 
> ```python
>     def getRunAsUserPrefixList(self, username=None):
>         if username is None:
>             username = getpass.getuser()
> 
>         if "SUDO" in list(os.environ.keys()):
>             return [os.environ["SUDO"], "-u", username]
>         else:
>             return ["sudo", "-u", username]
> ```
> 
> sudo is required here.

Can you file a bug against happy and put a TODO linking to it in CHIP ?

**Comment by mspang** [comment]
> > > > > Marking request changes for removal of root access requirement and use of `sudo` when running the tests via ninja. Please ping me if next steps are not clear.
> > > > 
> > > > 
> > > > PR Updated, found sudo is not available when using user namespace, added a face_sudo.sh as a workaround.
> > > 
> > > 
> > > Why is this required?
> > > From happy source:
> > > ```
> > >     def getRunAsRootPrefixList(self):
> > >         if "SUDO" in list(os.environ.keys()):
> > >             return [os.environ["SUDO"]]
> > >         elif os.getuid() == 0:
> > >             # We are already root
> > >             return []
> > >         else:
> > >             return ["sudo"]
> > > ```
> > > 
> > > 
> > > The `getuid()` check should work out, right?
> > 
> > 
> > In happy:
> > ```python
> >     def getRunAsUserPrefixList(self, username=None):
> >         if username is None:
> >             username = getpass.getuser()
> > 
> >         if "SUDO" in list(os.environ.keys()):
> >             return [os.environ["SUDO"], "-u", username]
> >         else:
> >             return ["sudo", "-u", username]
> > ```
> > 
> > 
> > sudo is required here.
> 
> Can you file a bug against happy and put a TODO linking to it in CHIP ?

Actually, I still don't think this is required. Looks like that function will not be run if we're already root.

**Comment by mspang** [comment]
> * Happy tests can be invoked by `RUN_HAPPY_TESTS=1 scripts/tests/gn_tests.sh`

This doesn't quite work, can you add some documentation before we add these to presubmit?

```
+ ninja -v -C scripts/tests/../../out/ -k 0 check
ninja: Entering directory `scripts/tests/../../out/'
ninja: error: loading 'build.ninja': No such file or directory
```

(These GitHub action scripts don't work well locally)

File: scripts/tests/happy_tests.sh [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+env
+
+function happytest_bootstrap() {
+    echo "Bootstrapping Happy Test"
+    set -e
+    # Bootstrap Happy
+    cd "$REPO_DIR/third_party/happy/repo"
+
+    # Override happy log dir config.
+    python3 -c 'import json, os; file = open("happy/conf/main_config.json"); data = json.load(file); data["state_file_prefix"] = os.environ.get("HAPPY_STATE_FILE_PREFIX", data["state_file_prefix"]); data["log_directory"] = os.environ["HAPPY_LOG_DIR"]; data["default_happy_log_dir"] = os.environ["HAPPY_LOG_DIR"]; out = open("happy/conf/main_config.json", "w"); json.dump(data, out);'
```
Comment (mspang) [inline scripts/tests/happy_tests.sh:38]: Can you explain why this is required?

If we run this as "root", which it seems we have to, then it writes a root owned filed to the checkout. Please don't do that.

Comment (erjiaqing) [reply scripts/tests/happy_tests.sh:38]: In CI, this is used to set the log path of happy test, by default, the logs are written to `/tmp`, and we want to set it to `/tmp/happy_test_logs`, the state file is set to avoid permission issue (can be found at https://github.com/erjiaqing/connectedhomeip/runs/1241935135).

overriting an existing file by root won't change the owner to root.

Comment (mspang) [reply scripts/tests/happy_tests.sh:38]: Please don't make edits to checked in files when building or testing. This will dirty everyone's checkout.

**Comment by erjiaqing** [comment]
> > * Happy tests can be invoked by `RUN_HAPPY_TESTS=1 scripts/tests/gn_tests.sh`
> 
> This doesn't quite work, can you add some documentation before we add these to presubmit?
> 
> ```
> + ninja -v -C scripts/tests/../../out/ -k 0 check
> ninja: Entering directory `scripts/tests/../../out/'
> ninja: error: loading 'build.ninja': No such file or directory
> ```
> 
> (These GitHub action scripts don't work well locally)

Added one more step for `gn gen`

File: scripts/tests/happy_tests.sh (hunk 13 above)
Comment (erjiaqing) [reply scripts/tests/happy_tests.sh]: After a update of happy, how happy tests for chip are using a dedicate `main_conf.json` in chip repo.

**Review by mspang — APPROVED** [review]

**Comment by mspang** [comment]
But the flow works reasonably well now and without extraordinary privileges, thank you.

I think we should followup with some further workflow fixes, e.g. it should work in parallel with the rest of the test suite and it should not print output to the console if it passes.

**Comment by mspang** [comment]
> I think we should followup with some further workflow fixes, e.g. it should work in parallel with the rest of the test suite and it should not print output to the console if it passes.

https://github.com/project-chip/connectedhomeip/issues/3214
